### PR TITLE
Enable deleting groups and members with authentication mapping group deletion

### LINKED ISSUE #27212: FE - Make SSO group mapping clearer
This is an update to the SSO group mapping UX to make the semantics clearer and (hopefully) resolve some recurring confusions about how group mapping works. The original bug motivating this project was https://github.com/metabase/metabase/issues/26079

**Links**
- product doc: [Make SSO group mapping clearer](https://www.notion.so/metabase/Make-SSO-group-mapping-clearer-b9ed33e4ee8c4facb2220f20b565e51f)
- figma link: [link](https://www.figma.com/file/Q75JpYNz6D109ZjoTGZ0ZP/Make-SSO-group-mapping-clearer?node-id=1%3A4&t=FeMud61cDfdX077F-0)

**Implementation Plan**
https://github.com/metabase/metabase/pull/26555 is a backend PR to add a new endpoint for clearing the members of a specific group. Once this is merged, the frontend can implement the changes described in the product doc and use the new endpoint for the `Remove mapping and members` option.

The actual update should be performed at the end when users click "Save". This means we should accumulate and correctly update a list of groups to be cleared/removed while editing mappings including corner cases of re-adding the same mapping, etc.

### PR DESCRIPTION
**Pull request by gusaiani**
Partially resolves https://github.com/metabase/metabase/issues/27212

### How to Test

📺 **Loom Demo**: https://www.loom.com/share/d90a32af70114d07bbf37b946e3b009b

1. In http://localhost:3000/admin/people/groups, create a few groups, and populate them with users
2. In http://localhost:3000/admin/settings/authentication/ldap, scroll all the way down and click on `Edit Mappings`
3. Create mappings, and populate them with zero to several groups each.
4. Save
5. Remove mappings
6. If a mapping is not linked to any groups, the mapping will simply disappear from the list. 
7. If a mapping is linked to one group, a modal will be displayed, with options on what to do about the groups in the mappings and copy using "group" in the singular from whenever appropriate.
8. If a mapping is linked to more than one group, a modal will be displayed, with options on what to do about the groups in the mappings and copy using "group" in the plural from whenever appropriate.

9. Save

Back in http://localhost:3000/admin/people/groups, you should:
1. no longer see the groups you had marked for deletion (Admin should still be there)
2. no longer see users in groups you had marked for emptying of users (Admin should still have the same users as before)

#### Corner cases 

1. If you create a mapping and click to Remove it before saving, no modal will appear. The groups mapped will not be affected.

### What will follow in upcoming PRs

1. The Mapping interaction will be inlined, removed from the Modal.
2. Unit tests for GroupMappingsWidget will be added.
3. Break-down GroupMappingsWidget.

### CODE REVIEW

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx [hunk 1]
```diff
@@ -10,34 +10,47 @@ import GroupSelect from "metabase/admin/people/components/GroupSelect";
...
       groups: null,
       mappings: {},
+      savedMappings: {},
       saveError: null,
+      dnForVisibleDeleteMappingModal: null,
+      groupIdsForVisibleDeleteMappingModal: null,
+      whenDeletingMappingGroups: {
+        groupsToClearAllPermissions: [],
+        groupsToDelete: [],
+      },
```
Comment (kulyk) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:36]: Doesn't seem like it's used while rendering. Should it live inside the component state?

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx [hunk 2]
```diff
@@ -77,27 +90,103 @@ export default class GroupMappingsWidget extends React.Component {
...
+  ) => {
+    const { groups } = this.state;
+
+    if (whatToDoAboutGroups === "nothing") {
+      return;
+    }
+
+    const allGroupIdsExceptAdmin = groupIds.filter(
+      groupId => !isAdminGroup(_.find(groups, group => group.id === groupId)),
+    );
```
Comment (kulyk) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:139]: Can we use `groups.find` instead of `_.find`?

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx [hunk 3]
```diff
@@ -108,9 +197,39 @@ export default class GroupMappingsWidget extends React.Component {
     );
   };
 
+  updateGroupsFromDeletedMappings = () => {
+    const { groupsToDelete } = this.state.whenDeletingMappingGroups;
+
+    groupsToDelete.forEach(id => PermissionsApi.deleteGroup({ id }));
```
Comment (alxnddr) [inline frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:203]: Call API directly from event handlers is not a great idea, it can also lead to bugs:

<video src="https://user-images.githubusercontent.com/14301985/214121809-a6590546-e631-40ef-9b9e-f6a4147e0892.mov" />

**Review by alxnddr — COMMENTED** [review]
I am wondering why we decided to go without inlining the form. Previously, we found out that it will lead to poor UX, and many edge cases and it is better to inline the form so the product doc and mockups got updated. Inlining the form removes that and makes the logic that tracks what to do with what groups unnecessary.

I'm asking because it seems that merging this PR will be like adding the new logic that should not exist and then removing it while implementing the project according to the spec

**Comment by kulyk** [comment]
@gusaiani @alxnddr, wanted to share some low-confidence feedback. While reviewing, I found it hard to follow the call sequence. Seems like there are a lot of possible things people can do about groups and mappings, hence a few levels of function calls). I have a feeling that the `GroupMappingsWidget` carries too many responsibilities. I wonder if it'd make sense to try to break it down? Maybe into different components or maybe extract a React hook encapsulating the business logic. I'm not so sure about that though as this is honestly the first time I'm looking at this code and it's almost my EoD 😅

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import Button from "metabase/core/components/Button";
+
+import {
+  ModalHeader,
+  ModalSubtitle,
+  ModalRadioRoot,
+} from "./DeleteGroupMappingModal.styled";
+
+export type GroupIds = number[];
+export type DNType = string;
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx:16]: Yes, 100% agreed, Kulyk.

This is coming from way above in the component tree.

Shall we do this in a separate PR?

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+export type ValueType = "nothing" | "clear" | "delete";
+
+type DeleteGroupMappingModalProps = {
+  dn: DNType;
+  groupIds: GroupIds;
+  onConfirm: (value: ValueType, groups: number[], dn: DNType) => void;
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx:22]: Right. Updated, thanks.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+}: DeleteGroupMappingModalProps) => {
+  const [value, setValue] = useState("nothing");
+
+  const handleChange = (newValue: ValueType) => {
+    setValue(newValue);
+  };
+
+  const handleConfirm = () => {
+    onConfirm(value as ValueType, groupIds, dn);
+  };
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx:40]: Ah, nice one. Thanks, updated.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.unit.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
+import React from "react";
+import { render, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+
+import type { GroupIds, DNType, ValueType } from ".";
+
+import DeleteGroupMappingModal from ".";
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.unit.spec.tsx:7]: 👍, updated

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.unit.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+type PropTypes = {
+  dn?: DNType;
+  groupIds?: GroupIds;
+  onConfirm?: (value: ValueType, groups: number[], dn: DNType) => void;
+  onHide?: () => void;
+};
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.unit.spec.tsx:14]: @kulyk, without some workaround, if we do that the calls to setup that pass props (for example in line 47) will beep.

As you can see this `PropTypes` is made up of optional `?` props.

Any tips on how to "have the cake and it eat it too", that is, using the imported PropTypes from the implementation file and having setup be ok with passing it fewer than all required props?

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx [hunk 9]
```diff
@@ -150,16 +269,37 @@ export default class GroupMappingsWidget extends React.Component {
...
-                      onDelete={this._deleteMapping(dn)}
-                    />
-                  ))}
+                  {Object.entries(mappings).map(([dn, ids]) => {
+                    const isMappingLinkedOnlyToAdminGroup =
+                      groups &&
+                      ids.length === 1 &&
+                      isAdminGroup(
+                        _.find(groups, group => group.id === ids[0]),
+                      );
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx]: Sure, updated.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx (hunk 2 above)
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx]: Sure, updated.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx (hunk 3 above)
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:203]: > I am wondering why we decided to go without inlining the form. 

@alxnddr let me try to recap the steps that made me go this way. 

It may be illustrative of how investing time in the transition from product doc to tickets (Github issues) can make us more productive, as our team talked about today in our retro.

### 1.

The issue that is on our tactical board, and is the gateway to this PR is https://github.com/metabase/metabase/issues/27212

Note how the issue description starts by saying "this is a small update".

Description is followed by links to the product doc and Figma. 
It may be worth noting that Figma is about the confirmation modal (what to do about groups), and afaik no Figma is about inlining the widget (understandable from one point-of-view, as it's intuitive what inlining would mean, but it brings the confirmation modal even more to the top as the main thing to tackle here).

Description next has "Implementation plan". 
This is what I went for first.
The implementation plan is what this PR does.
One thing that clearly needed to be tackled, and wasn't either in the implementation plan or in the product doc, was handling multiple groups per mapping, and edge cases around this fact. These were tackled here after discussions with Noah and Maz both in Slack and in Notion.

### 2.

After the above was taken care of, I went back to the product doc.

This PR was of a size that I already would veer towards doing further changes in a follow-up PR.

At any rate, I dedicated time to understanding different uses of the widget.
It became clear to me, and only at this point really, that this widget in SAML fits inside a narrow layout.
Also I learned that layouts for LDAP and JWT are different from the SAML layout.
Furthermore, the way the widget is called is a bit different in each of the three uses it gets.

I continued product discussions in our [Notion doc](https://www.notion.so/metabase/Make-SSO-group-mapping-clearer-b9ed33e4ee8c4facb2220f20b565e51f), see the top of the @today section.

And this discussion led to a decision by Maz and Arakaki on how to go about this.

**Comment by gusaiani** [comment]
> I wonder if it'd make sense to try to break it down

@kulyk, I fully agree.
As a matter of fact, this is in the PR description as a follow-up.

I think it will come hand-in-hand with extracting some logic to better places, as you point out as well.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx (hunk 3 above)
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:203]: From what's been decided today about inlining, what do you think of merging this iteration which works, and next we inline the widget?

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx (hunk 1 above)
Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:36]: Good call, thanks.

I'll check on answers to https://github.com/metabase/metabase/pull/27664#discussion_r1084698148 so I can tweak this, maybe we will inline the widget in this PR, or in a follow-up PR.

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.unit.spec.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.unit.spec.tsx:14]: @gusaiani you can use `Partial` like that:

```ts
type SetupOpts = Partial<DeleteGroupMappingModalProps>
```

It'd make every field optional. If you want to keep something mandatory, you can something like:

```ts
type Example = { a: string; b: number }

type PartialExample = Partial<Example> & {
  b: Example["b"
}

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget/DeleteGroupMappingModal/DeleteGroupMappingModal.tsx:16]: Sure, whatever makes the most sense 👍

**Comment by kulyk** [comment]
@gusaiani, my bad, I haven't noticed your comment about planned refactoring 🙈

File: frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx (hunk 3 above)
Comment (alxnddr) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:203]: @gusaiani got it, thanks for describing it comprehensively. It looks like we are slightly misaligned about what is the function of epics, issues, and product docs. This question has been raised a couple of times and as far as I remember we don't put full specifications with all the details in GitHub issues intentionally instead their role is more utilitarian and when necessary, providing additional info that can be valuable for engineers such as tech details, caveats. So before starting projects engineers need to read and understand product docs. If you think it makes sense to change the process and start putting the entire specs with all the details and mockups in the issues we need to align on that first. Something like an RFC on this process change would be helpful.

Regarding merging this approach as is, I think implementing the inline form right away would make more sense because we are going to ship the inline version and it has less logic. So the current implementation seems not like a first step towards the project implementation which would be fine to merge, but going in a different direction

Comment (gusaiani) [reply frontend/src/metabase/admin/settings/components/widgets/GroupMappingsWidget.jsx:203]: @alxnddr thanks.

I will continue this task by inlining the widget.
Will do this in a subbranch for now, just for a minor separation of diff states.

**Comment by gusaiani** [comment]
#28108 is doing the whole thing 👍
